src/MonoTorrent/MonoTorrent/InfoHash.cs

```diff
@@ -168,13 +168,28 @@ public static InfoHash FromHex (string infoHash)
             if (infoHash.Length != 40 && infoHash.Length != 64)
                 throw new ArgumentException ("V1 InfoHashes must be 40 characters long, V2 infohashes must be 64 characters long.", nameof (infoHash));
 
-            byte[] hash = new byte[infoHash.Length / 2];
-            for (int i = 0; i < hash.Length; i++)
-                hash[i] = byte.Parse (infoHash.Substring (i * 2, 2), System.Globalization.NumberStyles.HexNumber);
-
+            byte[] hash = HexStringToByteArray (infoHash);
             return InfoHash.FromMemory (hash);
         }
 
+        public static InfoHash FromMultiHash (string multiHash)
+        {
+            if (multiHash is null)
+                throw new ArgumentNullException (nameof (multiHash));
+
+            // the following may be too strict for 'truncated' sha-256 hashes which are allowed ??
+            if (multiHash.Length != 68)
+                throw new ArgumentException ("V2 multihashes must be 68 characters long.", nameof (multiHash));
```

Comment:
Ah I just realized truncated means it's a SHA1 sized (20 bytes) SHA2-256 variant, to have DHT lookup compatibility with SHA1. How would you prefer storing the truncated hash in Infohashes, extended with zeros?

Reply:
```
How would you prefer storing the truncated hash in Infohashes, extended with zeros?
```
Currently `InfoHash.Truncate ()` just returns the first 20 bytes of the underlying `Memory<byte>`. There's no distinction made between 20byte SHA1 and 20byte SHA256 info hashes at that point. They're (more or less) indistinguishable everywhere they're used.

When checking for equality, a truncated hash and full hash [will be considered equal](https://github.com/alanmcgovern/monotorrent/blob/1eff6105f75aba8fb35beaf9997aa2170b74ba9b/src/MonoTorrent/MonoTorrent/InfoHashes.cs#L108).

There is one place where I think a distinction will be critical: when somone creates an `InfoHash` object using a truncated SHA256 value, it *could* be mistakenly interpreted as a SHA1 value if you use `InfoHashes.FromInfoHash`

I wonder if having an enum as part of the InfoHash object would be beneficial.

```
public enum InfoHashType
{
    SHA1,
    SHA256
}
```

This would reduce the likelihood of confusing a 20 byte SHA1 hash with a 20 byte truncated SHA256 hash. When I started implementing this I wasn't (initially) sure that this would be necessary as I thought the full SHA256 hash would be available from MagnetLinks as well as when loading a .torrent file normally. However, if magnet links can be created using truncated hashes then the disambiguation would be helpful.

Reply:
Ah I took a look at the spec and code and decided that trying to track whether or not a specific InfoHash object is V1, V2  or truncated V2 is too messy.

There are (currently?) three ways to create a `TorrentManager`.
1. Use the .torrent file (or equivalent) and load it normally. At this point we have SHA1 and SHA256 infohashes so there's no confusion.
2. Use a MagnetLink. This has unambiguous V1 and V2 InfoHashes.
3. You have a raw infohash. If the infohash is 32 bytes long then it's clearly a V2 infohash, otherwise it's ambiguous and if you (the user of the library) don't know which it is, then there's no clear way for MonoTorrent to guess for you. In this case the simplest/safest option would be to use `ClientEngine.DownloadMetadataAsync`  to retrieve the metadata first, and then load that into the engine. WIth the actual metadata the full SHA1 and SHA256 infohashes can be computed and all is well. Additionally, if a tracker receives a 20 byte hash it can't tell if it's a V1 or V2, so at some point InfoHash objects (in monotorrent) will be ambiguous as they could represent either.

I retract that bad idea :P

Reply:
From inspecting the MetaDataMode.cs code, (when starting from a V2 only magnetlink) it looks like the DHT announce uses the truncated V2 hash, but the tracker `AnnounceRequest` and resulting `AnnounceMessage` are given the full length v2 hash, resulting in a invalid encoded announce message maybe?

Reply:
Yup! That's true, and that should be fixed now!